File: backend/src/application/dtos/RealStateDTO.ts
import 'reflect-metadata';
import {
  IsNotEmpty,
  IsString,
  IsNumber,
  IsEnum,
  IsBoolean,
  ValidateNested,
  ArrayMinSize,
  IsDefined,
  IsNotEmptyObject,
  IsObject,
  IsBooleanString,
} from "class-validator";


import { RealStateCategory } from "domain/types/types";
import { LocationDTO } from "./LocationDTO";
import { Transform, TransformFnParams, Type } from "class-transformer";

export class RealStateDTO {
  constructor(props: RealStateDTO) {
    Object.assign(this, props, { location: new LocationDTO(props.location)});
  }

  @IsNotEmpty({ message: "Title is required" })
  @IsString({ message: "Title must be a string" })
  title: String;

  @IsNotEmpty({ message: "Price is required" })
  @IsNumber({}, { message: "Price must be a Number" })
  price: Number;


  @IsNotEmpty({ message: "Category is required" })
  @IsEnum(RealStateCategory, { message: "Invalid category" })
  category: RealStateCategory;

  // @IsNotEmpty({ message: "Rooms is required" })
  // @IsNumber({}, { message: "Rooms must be a Number" })
  rooms: Number | null;

  // @IsNotEmpty({ message: "Beds is required" })
  // @IsNumber({}, { message: "Beds must be a Number" })
  beds: Number | null;

  // @IsNotEmpty({ message: "Baths is required" })
  // @IsNumber({}, { message: "Baths must be a Number" })
  baths: Number | null;

  // @IsNotEmpty({ message: "Kitchens is required" })
  // @IsNumber({}, { message: "Kitchens must be a Number" })
  kitchens: Number | null;

  seats: Number | null;

  @IsNotEmpty({ message: "Size is required" })
  @IsNumber({}, { message: "widthSize must be a Number" })
  sizeWidth: Number;

  @IsNotEmpty({ message: "Size is required" })
  @IsNumber({}, { message: "heightSize must be a Number" })
  sizeHeight: Number;

  @IsNotEmpty({ message: "Size unit is required" })
  @IsString({ message: "Size unit must be a string" })
  sizeUnit: String;

  @IsObject()
  @IsNotEmptyObject()
  @ValidateNested()
  @Type(() => LocationDTO)
  location: LocationDTO;

  otherFeatures: String[];

  @IsNotEmpty({ message: "Description is required" })
  @IsString({ message: "Description must be a string" })
  description: String;

  @Transform(({ value }: TransformFnParams) => value?.trim())
  @IsBoolean({ message: "isApproved must be a boolean"})
  isApproved: Boolean;

  @IsNotEmpty({ message: "Owner id is required" })
  @IsString({ message: "Owner must be a string" })
  owner: String;

  @IsNotEmpty({ message: "number of views are required" })
  @IsNumber({}, { message: "number of views must be a Number" })
  numberOfViews: Number

  status: String | null;
}

File: mobile/dalali_hub/lib/data/remote/model/realstate_dto.dart
import 'dart:io';

import 'package:dalali_hub/data/remote/model/location_dto.dart';
import 'package:dalali_hub/domain/entity/realstate.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'realstate_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class RealstateDto {
 final String? id;
  final String title;
  final double price;
  final double? rooms;
  final double? beds;
  final double? baths;
  final double? kitchens;
  final double sizeWidth;
  final double sizeHeight;
  final String sizeUnit;
  final List<String> otherFeatures;
  final String description;
  final bool isApproved;
  final String category;
  final Map location;
  final int? seats;
  final int numberOfViews;
  

  RealstateDto({
    required this.title,
    required this.category,
    required this.price,
    this.rooms,
    this.beds,
    this.baths,
    this.kitchens,
    required this.sizeWidth,
    required this.sizeHeight,
    required this.sizeUnit,
    required this.otherFeatures,
    required this.description,
    required this.isApproved,
    this.id,
    required this.location,
    this.seats,
    required this.numberOfViews
  });
  Map<String, dynamic> toJson() => _$RealstateDtoToJson(this);

 static RealstateDto toRealstateDto(Realstate realstate) {
    return RealstateDto(
      id: realstate.id,
      title: realstate.title,
      category: realstate.category,
      price: realstate.price,
      rooms: realstate.rooms,
      beds: realstate.beds,
      baths: realstate.baths,
      kitchens: realstate.kitchens,
      sizeWidth: realstate.sizeWidth,
      sizeHeight: realstate.sizeHeight,
      sizeUnit: realstate.sizeUnit,
      otherFeatures: realstate.otherFeatures,
      description: realstate.description,
      isApproved: realstate.isApproved,
      location: {
        "region": realstate.location.region,
        "district": realstate.location.district,
        "ward": realstate.location.ward
      },
      numberOfViews: realstate.numberOfViews,
      
      
    );
  }
}
